drashland/rhum
# chore: DRY code; separate logic into files where needed; add comments

Closes #153

## Summary
- First attempt to DRY code. The Spy PR #162 is going to be branched off this so that it doesn't further introduce duplicated code.
- Adds `TestDoubleBuilder` that mocks and fakes can extend -- sharing the same logic to help them build themselves as test doubles
- Separates `Error` classes into `errors.ts` file
- Adds `MethodVerificationError` to show a better stack trace when methods being verified throw errors (inspiration taken from the test runner we were going to make for v2):
  ```text
    test .willThrow() causes throwing RandomError (with constructor) ... ok (4ms)
    test .willThrow() causes throwing RandomError2 (no constructor) ... ok (2ms)
    test .expects(...).toBeCalled(...) ... FAILED (4ms)


      MethodVerificationError: Method "hello" received incorrect number of calls.
          at fn (file:///drashland/rhum/tests/deno/unit/mod/mock_test.ts:250:14)

      Verification Results:
          Expected calls -> 1
          Actual calls   -> 2

      Check the above 'mock_test.ts' file at/around line 250 for the following code to fix this error:
          .method("hello").toBeCalled(1)


    FAILED (29ms)
    test call count for outside nested function is increased ... ok (3ms)
  ```
- Add hella docblocks with example code blocks
- Remove dead code (types that aren't being used and such)
